Mathlib/Data/Pi/Algebra.lean

```diff
@@ -0,0 +1,429 @@
+theorem mulSingle_comm [One β] (i : I) (x : β) (i' : I) :
+    (mulSingle i x : I → β) i' = (mulSingle i' x : I → β) i := by
+  simp [mulSingle_apply, eq_comm]
+
+@[to_additive]
+theorem apply_mulSingle (f' : ∀ i, f i → g i) (hf' : ∀ i, f' i 1 = 1) (i : I) (x : f i) (j : I) :
+    f' j (mulSingle i x j) = mulSingle i (f' i x) j := by
+  simpa only [Pi.one_apply, hf', mulSingle] using Function.apply_update f' 1 i x j
+
+@[to_additive apply_single₂]
+theorem apply_mulSingle₂ (f' : ∀ i, f i → g i → h i) (hf' : ∀ i, f' i 1 1 = 1) (i : I)
+    (x : f i) (y : g i) (j : I) :
+    f' j (mulSingle i x j) (mulSingle i y j) = mulSingle i (f' i x y) j := by
+  by_cases h : j = i
+  · subst h
+    simp only [mulSingle_eq_same]
+
+  · simp only [mulSingle_eq_of_ne h, hf']
+
+@[to_additive]
+theorem mulSingle_op {g : I → Type _} [∀ i, One (g i)] (op : ∀ i, f i → g i)
+    (h : ∀ i, op i 1 = 1) (i : I) (x : f i) :
+    mulSingle i (op i x) = fun j => op j (mulSingle i x j) :=
+  Eq.symm <| funext <| apply_mulSingle op h i x
+
+@[to_additive]
+theorem mulSingle_op₂ {g₁ g₂ : I → Type _} [∀ i, One (g₁ i)] [∀ i, One (g₂ i)]
+    (op : ∀ i, g₁ i → g₂ i → f i) (h : ∀ i, op i 1 1 = 1) (i : I) (x₁ : g₁ i) (x₂ : g₂ i) :
+    mulSingle i (op i x₁ x₂) = fun j => op j (mulSingle i x₁ j) (mulSingle i x₂ j) :=
+  Eq.symm <| funext <| apply_mulSingle₂ op h i x₁ x₂
+
+variable (f)
+
+@[to_additive]
+theorem mulSingle_injective (i : I) : Function.Injective (mulSingle i : f i → ∀ i, f i) :=
+  Function.update_injective _ i
+
+@[simp, to_additive]
+theorem mulSingle_inj (i : I) {x y : f i} : mulSingle i x = mulSingle i y ↔ x = y :=
+  (Pi.mulSingle_injective _ _).eq_iff
+
+#align pi.mul_single_eq_same Pi.mulSingle_eq_same
+#align pi.mul_single_eq_of_ne Pi.mulSingle_eq_of_ne
+#align pi.mul_single_eq_of_ne' Pi.mulSingle_eq_of_ne'
+#align pi.mul_single_one Pi.mulSingle_one
+#align pi.mul_single_apply Pi.mulSingle_apply
+#align pi.mul_single_comm Pi.mulSingle_comm
+#align pi.apply_mul_single Pi.apply_mulSingle
+#align pi.apply_mul_single₂ Pi.apply_mulSingle₂
+#align pi.mul_single_op Pi.mulSingle_op
+#align pi.mul_single_op₂ Pi.mulSingle_op₂
+#align pi.mul_single_injective Pi.mulSingle_injective
+#align pi.mul_single_inj Pi.mulSingle_inj
+
+end
+
+/-- The mapping into a product type built from maps into each component. -/
+@[simp]
+protected def prod (f' : ∀ i, f i) (g' : ∀ i, g i) (i : I) : f i × g i :=
+  (f' i, g' i)
+
+-- Porting note : These two are simp-normal-form versions of the lemmas below.
+-- They should probably be placed in a lower file and be renamed.
+@[simp]
+theorem prod_fst_snd₂ : (fun (i : α × β) => i) = id :=
+  funext fun _ => Prod.mk.eta
+
+-- Porting note : These two are simp-normal-form versions of the lemmas below.
+-- They should probably be placed in a lower file and be renamed.
+@[simp]
+theorem prod_snd_fst₂ : (fun (i : α × β) => (i.snd, i.fst)) = Prod.swap :=
+  rfl
```

Comment:
I think just delete these, and on the next two theorems write:
`Porting note: simp now unfolds the lhs, so we are not marking these as simp`.
and hope for the best?